Returning comptime if gives builder error

### Describe the bug

.

### Expected Behavior

It should compile.

### Current Behavior

```
/tmp/v_1000/compime_ret.4290671977799971669.tmp.c:12133: error: cannot convert 'struct string' to 'struct _result_string'
builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

```v
fn comptime_ret_result() !string {
	return $if debug {
		'debug'
	} $else {
		'not debug'
	}
}

println(comptime_ret_result()!)
```

### Possible Solution

_No response_

### Additional Information/Context

Returning them inside the if block works.

```v
fn comptime_ret_result() !string {
	$if debug {
		return 'debug'
	} $else {
		return 'not debug'
	}
}

println(comptime_ret_result()!)
```

Running it with `v run file.v` gives the following output:

```
not debug
```

### V version

V 0.3.2 a22dfe0

### Environment details (OS name and version, etc.)

OS: linux, "Arch Linux"
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-4570 CPU @ 3.20GHz
CC version: cc (GCC) 12.2.0

vexe mtime: 2023-01-17 15:46:59
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.2 2034dcb.a22dfe0

Git version: git version 2.39.0
Git vroot status: weekly.2023.03-8-ga22dfe0d
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/gen/c/comptime.v
+++ b/vlib/v/gen/c/comptime.v
@@ -285,6 +285,7 @@ fn (mut g Gen) comptime_if(node ast.IfExpr) {
 		}
 	}
 	tmp_var := g.new_tmp_var()
+	is_opt_or_result := node.typ.has_flag(.option) || node.typ.has_flag(.result)
 	line := if node.is_expr {
 		stmt_str := g.go_before_stmt(0)
 		g.write(util.tabs(g.indent))
@@ -339,8 +340,16 @@ fn (mut g Gen) comptime_if(node ast.IfExpr) {
 					g.set_current_pos_as_last_stmt_pos()
 					prev_skip_stmt_pos := g.skip_stmt_pos
 					g.skip_stmt_pos = true
-					g.write('\t${tmp_var} = ')
-					g.stmt(last)
+					if is_opt_or_result {
+						tmp_var2 := g.new_tmp_var()
+						g.write('{ ${g.base_type(node.typ)} ${tmp_var2} = ')
+						g.stmt(last)
+						g.writeln('_result_ok(&(${g.base_type(node.typ)}[]) { ${tmp_var2} }, (_result*)(&${tmp_var}), sizeof(${g.base_type(node.typ)}));')
+						g.writeln('}')
+					} else {
+						g.write('\t${tmp_var} = ')
+						g.stmt(last)
+					}
 					g.skip_stmt_pos = prev_skip_stmt_pos
 					g.writeln(';')
 					g.writeln('}')
@@ -350,8 +359,16 @@ fn (mut g Gen) comptime_if(node ast.IfExpr) {
 					g.set_current_pos_as_last_stmt_pos()
 					prev_skip_stmt_pos := g.skip_stmt_pos
 					g.skip_stmt_pos = true
-					g.write('${tmp_var} = ')
-					g.stmt(last)
+					if is_opt_or_result {
+						tmp_var2 := g.new_tmp_var()
+						g.write('{ ${g.base_type(node.typ)} ${tmp_var2} = ')
+						g.stmt(last)
+						g.writeln('_result_ok(&(${g.base_type(node.typ)}[]) { ${tmp_var2} }, (_result*)(&${tmp_var}), sizeof(${g.base_type(node.typ)}));')
+						g.writeln('}')
+					} else {
+						g.write('${tmp_var} = ')
+						g.stmt(last)
+					}
 					g.skip_stmt_pos = prev_skip_stmt_pos
 					g.writeln(';')
 					g.indent--

--- a/vlib/v/tests/ifexpr_with_option_result_test.v
+++ b/vlib/v/tests/ifexpr_with_option_result_test.v
@@ -1,0 +1,11 @@
+fn comptime_ret_result() !string {
+	return $if debug {
+		'debug'
+	} $else {
+		'not debug'
+	}
+}
+
+fn test_main() {
+	assert comptime_ret_result()? == 'not debug'
+}

